### PR TITLE
Allow disabling character set filtering

### PR DESCRIPTION
**Pull request by DemiMarie**
This allows copying paths with names that would otherwise be forbidden.
This requires adding new APIs, so take the opportunity to provide more
useful information about why a path was rejected.

To ensure that internal invariants are not violated, this uses a new
COMPILETIME_UNREACHABLE macro to validate that a statement can be proven
unreachable by the compiler.  This is superior to abort() because it is
checked at compile-time by the optimizer: if the compiler cannot prove
that the code is unreachable, the build will fail.  This technique is
also used by BUILD_BUG_ON() in the Linux kernel.  Since compilers do not
promise to always be able to prove a piece of code unreachable, static
checking is disabled by default.  It can be enabled by including
CHECK_UNREACHABLE=1 in the build environment.

This is part of QubesOS/qubes-issues#8332 (less restricted qfile-copy),
but that issue also requires changes to qfile-unpacker (part of
qubes-core-agent-linux) to fix.  The code is both backwards and forwards
compatible: Old qfile-unpacker versions will work fine with the new
library, and new qfile-unpacker versions will work fine with the old
library.  However, qubes.UnsafeFileCopy will behave like qubes.Filecopy
unless the library has been updated.

Initially, I decided to unconditionally disallow ASCII control
characters in filenames, even if other character set filtering is
disabled.  This is because they are very useful for exploits and not
very useful for other purposes.  However, they can still arise in
practice for completely legitimate reasons.  These include copying and
pasting into a file name in a GUI file manager and deliberately creating
strangely-named files for test purposes.  Therefore, disabling filtering
now disables _all_ character set filtering.  Restrictions to prevent
directory traversal are still enforced, though, because violations of
these are much more likely to be exploitable.  While symlinks that point
outside of the directory being copied might arise legitimately, they
will not be meaningful after being copied and _do_ allow an attacker to
cause mischief.  For instance, if a symlink points to
/home/user/.bashrc, "cp a b" in a directory copied by qvm-copy could
overwrite ~/.bashrc with attacker-controlled data.  The checks on
symlink paths prevent this.

### CODE REVIEW

**Comment by marmarek** [comment]
> While symlinks that point
> outside of the directory being copied might arise legitimately, they
> will not be meaningful after being copied and _do_ allow an attacker to
> cause mischief

See task description - please add a flag to disable this part too (do expose it as an option for `qfile-unpacker` but do _not_ enable it for `qubes.UnsafeFilecopy`).

**Comment by DemiMarie** [comment]
> > While symlinks that point
> > outside of the directory being copied might arise legitimately, they
> > will not be meaningful after being copied and _do_ allow an attacker to
> > cause mischief
> 
> See task description - please add a flag to disable this part too (do expose it as an option for `qfile-unpacker` but do _not_ enable it for `qubes.UnsafeFilecopy`).

Will do.  I’ll keep validating the paths from the source VM, though.  These are produced by `qfile-agent`, so they will be well-formed no matter what is being copied.

File: qrexec-lib/validator-test.c [hunk 1]
```diff
@@ -70,7 +72,137 @@ int main(int argc, char **argv)
...
+        TEST("a/b", "../a", 0, false),
+        TEST("a/b", "../a/b/c", 0, false),
+        TEST("a/b", "../a/b/c", 0, false),
+        TEST("a/b/c", "../../a", 0, false),
+        TEST("a/b/c", "../../a", 0, false),
+        TEST("a/b", "a", 0, true),
+        TEST("a/b/c", "../a", 0, true),
+        // Absolute symlinks are rejected
+        TEST("a/b/c", "/a", 0, false),
+        TEST("a/b/c", "/a", 0, false),
```
Comment (marmarek) [inline qrexec-lib/validator-test.c:168]: A bunch of tests are duplicated here.

File: qrexec-lib/validator-test.c [hunk 2]
```diff
@@ -70,7 +72,137 @@ int main(int argc, char **argv)
...
+        TEST("a/b/c", "a/", 0, true),
+        // Invalid paths are rejected...
+        TEST("..", "a/", 0, false),
+        // ...even with QUBES_PURE_ALLOW_UNSAFE_SYMLINKS.
+        TEST("..", "a/", QUBES_PURE_ALLOW_UNSAFE_SYMLINKS, false),
+        // but QUBES_PURE_ALLOW_UNSAFE_SYMLINKS allows bad symlinks
+        TEST("a", "/bin/sh", QUBES_PURE_ALLOW_UNSAFE_SYMLINKS, true),
+        // QUBES_PURE_ALLOW_NON_CANONICAL_SYMLINKS allows non-canonical symlinks...
+        TEST("a/b", "b//c", QUBES_PURE_ALLOW_NON_CANONICAL_SYMLINKS, true),
+        TEST("a/b", "b/./c", QUBES_PURE_ALLOW_NON_CANONICAL_SYMLINKS, true),
```
Comment (marmarek) [inline qrexec-lib/validator-test.c]: If you check if stuff like this is allowed with the flag, shouldn't you check also if it's refused without it?

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
> Installing updates failed, skipping the report!

```
qfile-agent: Fatal error: File copy: "Invalid or incomplete multibyte or wide character; Last file: adobe-mappings-cmap-20230622-1.fc37.noarch.rpm" (error type: Invalid or incomplete multibyte or wide character)
```

Doesn't look like a name that should be rejected in any mode,,,

**Comment by marmarek** [comment]
And while at it, can you change the error message for EILSEQ to something clearer for mere mortals? Maybe "File name contains disallowed characters"?

**Comment by DemiMarie** [comment]
> > Installing updates failed, skipping the report!
> 
> ```
> qfile-agent: Fatal error: File copy: "Invalid or incomplete multibyte or wide character; Last file: adobe-mappings-cmap-20230622-1.fc37.noarch.rpm" (error type: Invalid or incomplete multibyte or wide character)
> ```
> 
> Doesn't look like a name that should be rejected in any mode,,,

It is accepted by `qubes_pure_validate_file_name()`.  I will need to do an integration test to see what happens.

**Comment by DemiMarie** [comment]
> And while at it, can you change the error message for EILSEQ to something clearer for mere mortals? Maybe "File name contains disallowed characters"?

Already done in afade05e2a38a01b59d2a02da36faada5ebdb4fe, but EILSEQ is also used for non-canonical symbolic links.

**Comment by marmarek** [comment]
Another line from logs (but doesn't really add here):
```
Command '['/usr/libexec/qubes/qfile-dom0-unpacker', '1000', '/var/tmp/qubes-updates-tmp3tzhg9ms.UNTRUSTED', '--only-regular-files']' returned non-zero exit status 84.
```

**Comment by DemiMarie** [comment]
> > Installing updates failed, skipping the report!
> 
> ```
> qfile-agent: Fatal error: File copy: "Invalid or incomplete multibyte or wide character; Last file: adobe-mappings-cmap-20230622-1.fc37.noarch.rpm" (error type: Invalid or incomplete multibyte or wide character)
> ```
> 
> Doesn't look like a name that should be rejected in any mode,,,

Oops, I switched to the `*_v2` functions in acc900a522bcb26e214c5aa4f00bd5a9324791ee, but forgot to change the return value checks.  The new functions return an `int` (0 on success, negative on failure), while the old functions returned a `bool` (true on success, false on failure), so the result was that a file was accepted if it should be rejected and visa versa.  Fixed in 8c17abd2e5ff57e38af8a612e7578fc36c1da599.

**Comment by marmarek** [comment]
This (or the other PR) broke salt, specifically the step when files from dom0 are copied into dispvm:
```
[2024-05-20 01:09:52] [   18.208751] qrexec-agent[625]: 2024-05-20 01:09:52.950 qrexec-agent[625]: qrexec-agent-data.c:288:handle_new_process_common: pid 627 exited with 95
```

I don't have stderr logged to see what file it complained about, but I'll try to save it next time.

**Comment by marmarek** [comment]
I got the stderr, and I'm more confused:
```
File copy: "Operation not supported; Last file:  srv/formulas/test/update-formula/update"
```
The path here is a normal directory.

**Comment by marmarek** [comment]
And relevant part of strace on the unpacker:
```
738   read(0, "srv/formulas/test/update-formula"..., 40) = 40
738   fcntl(0, F_GETFL)                 = 0x2 (flags O_RDWR)
738   openat(AT_FDCWD, "srv", O_RDONLY|O_NOCTTY|O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY) = 4
738   openat(4, "formulas", O_RDONLY|O_NOCTTY|O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY) = 5
738   close(4)                          = 0
738   openat(5, "test", O_RDONLY|O_NOCTTY|O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY) = 4
738   close(5)                          = 0
738   openat(4, "update-formula", O_RDONLY|O_NOCTTY|O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY) = 5
738   close(4)                          = 0
738   mkdirat(5, "update", 0700)        = -1 EEXIST (File exists)
738   newfstatat(AT_FDCWD, "srv/formulas/test/update-formula/update", {st_mode=S_IFDIR|0700, st_size=4096, ...}, 0) = 0
738   openat(5, "update", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 4
738   newfstatat(4, "", {st_mode=S_IFDIR|0700, st_size=4096, ...}, AT_EMPTY_PATH) = 0
738   chmod("/proc/self/fd/4", 0755)    = -1 ENOENT (No such file or directory)
738   close(4)                          = 0
738   close(0)                          = 0
738   write(1, "_\0\0\0\0\0\0\0\\m\26v\0\0\0\0", 16) = 16
738   write(1, "'\0\0\0", 4)            = 4
738   write(1, "srv/formulas/test/update-formula"..., 39) = 39
738   exit_group(95)                    = ?
738   +++ exited with 95 +++
```

**Comment by marmarek** [comment]
> 738   chmod("/proc/self/fd/4", 0755)    = -1 ENOENT (No such file or directory)

This is most likely due to chroot. There is saved open procfd_fd, but this particular call is likely done by glibc (`fchmodat` with `AT_SYMLINK_NOFOLLOW`?) so it doesn't know how to use it.

File: qrexec-lib/unpack.c [hunk 3]
```diff
@@ -154,7 +154,7 @@ static void fix_times_and_perms(const int fd,
     } else {
         /* Do not change the mode of what a symbolic link points to */
         if (!S_ISLNK(untrusted_hdr->mode) &&
-            fchmodat(fd, last_segment, untrusted_hdr->mode & 07777, 0))
+            fchmodat(fd, last_segment, untrusted_hdr->mode & 07777, AT_SYMLINK_NOFOLLOW))
```
Comment (marmarek) [inline qrexec-lib/unpack.c]: checking glibc git log says:
```
commit 65341f7bbea824d2ff9d37db15d8be162df42bd3
Author: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date:   Thu Sep 28 13:56:21 2023 -0300

    linux: Use fchmodat2 on fchmod for flags different than 0 (BZ 26401)
    
    Linux 6.6 (09da082b07bbae1c) added support for fchmodat2, which has
    similar semantics as fchmodat with an extra flag argument.  This
    allows fchmodat to implement AT_SYMLINK_NOFOLLOW and AT_EMPTY_PATH
    without the need for procfs.
    
    The syscall is registered on all architectures (with value of 452
    except on alpha which is 562, commit 78252deb023cf087).
    
    The tst-lchmod.c requires a small fix where fchmodat checks two
    contradictory assertions ('(st.st_mode & 0777) == 2' and
    '(st.st_mode & 0777) == 3').
```

Linux 6.6 (and glibc 2.39) is way too new to rely on this feature, and the fallback via procfs can't work here. Drop the change.

Comment (DemiMarie) [reply qrexec-lib/unpack.c]: Reimplemented as 075dfe462639f7d9a205138558cbf5a3a64d2366 using `fchmod()` and moved to a separate PR (#114).

File: qrexec-lib/unicode.c [hunk 4]
```diff
@@ -101,25 +102,99 @@ static int validate_utf8_char(const uint8_t *untrusted_c) {
...
+// and a bind mount, but the bind mount is not always effective if mount
+// namespaces are in use, and the chroot can be bypassed (QSB-015).
+//
+// Preconditions:
+//
+// - untrusted_name is NUL-terminated.
+// - allowed_leading_dotdot is the maximum number of leading "../" sequences
+//   allowed.  Might be 0.
+// - allow_unsafe_chars is true if and only if unsafe characters should _not_
+//   be allowed.
```
Comment (marmarek) [inline qrexec-lib/unicode.c]: Isn't the comment the other way around? But also, there is no `allow_unsafe_chars`, but `flags` instead.

File: qrexec-lib/pack.c [hunk 5]
```diff
@@ -107,6 +107,18 @@ void wait_for_result(void)
             case EINVAL:
                 call_error_handler("File copy: Corrupted data from packer%s\"%s\"", last_filename_prefix, last_filename);
                 break;
+            case EILSEQ:
+                // FIXME: the sanitized filename is not useful here, but the unsanitized filename
+                // is dangerous (which is likely why it was rejected in the first place!).
+                call_error_handler("Forbidden character in file or link target");
```
Comment (marmarek) [inline qrexec-lib/pack.c]: Still, some hint which name it is would be useful. Sanitized name will be distorted a bit (say that in the message?), but still can help to find the issue.

**Review by marmarek — CHANGES_REQUESTED** [review]

**Review by marmarek — APPROVED** [review]

**Pull request merged**